ostd/src/sync/rcu/mod.rs

```diff
@@ -204,24 +200,35 @@ impl<P: OwnerPtr> Drop for RcuInner<P> {
 }
 
 /// The inner implementation of both [`RcuReadGuard`] and [`RcuOptionReadGuard`].
-struct RcuReadGuardInner<'a, P: OwnerPtr> {
-    obj_ptr: *mut <P as OwnerPtr>::Target,
+struct RcuReadGuardInner<'a, P: NonNullPtr> {
+    ptr_ref: Option<P::Ref<'a>>,
```

Comment:
This still needs to be `*mut Target`.

The convertion to `Ref<'_>` needs to happen latter to avoid the soundness issue (as we cannot really have `Ref<'a>`):
```rust
impl RcuReadGuardInner {
    pub fn get(&self) -> Option<&P::Ref<'_>> { .. }
}
```